cjwkernel/pandas/types.py

```diff
@@ -945,192 +926,125 @@ def coerce(
+            columns = _infer_columns(dataframe, {}, try_fallback_columns)
+            return cls(dataframe=dataframe, errors=errors)
+        else:
+            raise ValueError(
+                "Expected (Dataframe, RenderError) or (str, dict) return type; got (%s,%s)"
+                % (type(value[0]).__name__, type(value[1]).__name__)
             )
 
-        return cls(
-            errors=[
-                ProcessResultError.module_exception(
-                    "Invalid return type %s" % type(value).__name__
+    @classmethod
+    def _coerce_3tuple(
+        cls, value, try_fallback_columns: Iterable[Column] = []
+    ) -> ProcessResult:
+        dataframe, error, json = value
+        if dataframe is None:
+            dataframe = pd.DataFrame()
+        elif not isinstance(dataframe, pd.DataFrame):
+            raise ValueError("Expected DataFrame got %s" % type(dataframe).__name__)
+        if json is None:
+            json = {}
+        elif not isinstance(json, dict):
+            raise ValueError("Expected JSON dict, got %s" % type(json).__name__)
+
+        errors = ProcessResultError.coerce_list(error)
+
+        validate_dataframe(dataframe)
+        columns = _infer_columns(dataframe, {}, try_fallback_columns)
+        return cls(dataframe=dataframe, errors=errors, json=json, columns=columns)
+
+    @classmethod
+    def _coerce_dict(
+        cls, value, try_fallback_columns: Iterable[Column] = []
+    ) -> ProcessResult:
+        if "message" in value and "quickFixes" in value:
+            return cls(errors=[ProcessResultError.coerce(value)])
+        else:
+            value = dict(value)  # shallow copy
+            try:
+                errors = ProcessResultError.coerce_list(value.pop("errors"))
+            except KeyError:
+                errors = []
+            except ValueError as e:
+                raise ValueError("Invalid 'errors' property") from e
+
+            # Coerce old-style error and quick_fixes, if it's there
+            try:
+                legacy_error_message = I18nMessage.coerce(value.pop("error"))
+            except KeyError:
+                legacy_error_message = None
+            except ValueError as e:
+                raise ValueError("Invalid 'error' property") from e
+
+            try:
+                legacy_error_quick_fixes = [
+                    QuickFix.coerce(v) for v in value.pop("quick_fixes")
+                ]
+            except KeyError:
+                legacy_error_quick_fixes = []
+
+            if legacy_error_quick_fixes and not legacy_error_message:
+                raise ValueError("You cannot return quick fixes without an error")
+            if legacy_error_message:
+                errors.append(
+                    ProcessResultError(legacy_error_message, legacy_error_quick_fixes)
                 )
-            ]
-        )
+
+            dataframe = value.pop("dataframe", pd.DataFrame())
+            validate_dataframe(dataframe)
+
+            try:
```

Comment:
With the indentation fixed, I can see another bug now: this doesn't always call `_infer_columns()`.

The fix: nix try/catch:

```
column_formats = value.pop("column_formats", {})
value["columns"] = _infer_columns(dataframe, column_formats, try_fallback_columns)
```

Reply:
So `column_formats` becomes required?